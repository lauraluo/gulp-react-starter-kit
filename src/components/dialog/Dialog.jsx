import Reflux from "reflux"
import React from 'react'
import $ from "jquery"
import DialogButton from "./DialogButton"
import DialogRelayoutMixin from "../core/Mixin.DialogRelayout"

var DialogActions = Reflux.createActions([
    "showDialog",
    "showDialogYesNo",
    "showDialogList",
    "showDialogListYesNo",
    "showDialogScene",
    "showDialogToast",
    "hideDialog",
    "updateDialog",
    "setDialogName",
    "clearDialogName"
]);

class DialogStore extends Reflux.Store {
    constructor(props) {
        super(props);
        this.getInitState = function() { 
            return {
                isOpened: false,
                posY: 0,
                type: 0,
                title: '',
                content: {},
                buttons: [],
                didOpened: {},
                willOpen: {},
                didClosed: {}
            }
        };
        this.state = this.getInitState();
        this.listenables = DialogActions;
    }

    _updateState() {
        if (!this.state.isOpened && this.state.type != 0) {
            this.state.posY = $(document).scrollTop();
        }

        this.state.isOpened = (this.state.type != 0);
        this.trigger(this.state);
    }

    _checkConfigsInterface = (configs) => {
        var result = true;

        Object
            .keys(configs)
            .map((key, index) => {
                var value = configs[key];
            
                switch (key) {
                    case "title":
                        if (typeof value !== "string") {
                            console.error("title should be a string");
                            result = false;
                        }
                        break;
                    case "content":
                        if (typeof value !== "function" && typeof value !== "string") {
                            console.error("content should be a functional component or a string");
                            result = false;
                        }
                        break;
                    case "buttons":
                        if (Object.prototype.toString.call(value) !== '[object Array]') {
                            console.error("buttons should be an Array");
                            result = false;
                        }
                        break;
                    case "didOpened":
                    case "willOpen":
                    case "didClosed":
                        if (value === {} && typeof value !== "function") {
                            console.error("callback should be an function");
                            result = false;
                        }
                        break;
                    default:
                        return;
                }
            });

        return result;

    }

    onShowDialog = (inputConfigs) => {

        var configs = Object.assign(this.getInitState(),inputConfigs);

        if (!this._checkConfigsInterface(configs)) {
            return;
        }

        this.state.type = 1;
        this.state.title = configs.title;
        this.state.content = configs.content;
        this.state.buttons = configs.buttons;
        this.state.didOpened = configs.didOpened;
        this._updateState();
    }

    onHideDialog = () => {
        this.state = Object.assign({},this.getInitState());
        this._updateState();
    }

    onUpdateDialog = (updateConfigs) => {
        if(updateConfigs.title || updateConfigs.title === '') this.state.title = updateConfigs.title;
        if(updateConfigs.content)  this.state.content = updateConfigs.content;
        if(updateConfigs.buttons) this.state.buttons = updateConfigs.buttons;
        this._updateState();
    }
}

class Dialog extends Reflux.PureComponent {
    constructor(props) {
        super(props);
        this.state = {}; // our store will add its own state to the component's
        this.store = DialogStore; // <- just assign the store class itself
        this._BodyComponent = this
            ._BodyComponent
            .bind(this);
        this._relayout = this
            .props
            ._relayout
            .bind(this);
    }

    componentWillUpdate(nextProps, nextState) {}

    componentDidUpdate(prevProps, prevState) {
        if (this.state.isOpened && !prevState.isOpened && typeof this.state.didOpened == "function") {
            this.state.didOpened();
        }
        this._relayout();
    }

    _BodyComponent = (props) => {
        var className = (props.extraPadding)
            ? 'dialog-box-body dialog-box-body-padding'
            : 'dialog-box-body';
        var renderResult;

        if (typeof(this.state.content) === 'string') {
            let content = this
                .state
                .content
                .split('\n')
                .map(function (item, i) {
                    return (
                        <span key={'dialog-item' + i}>{item}<br/></span>
                    );
                });

            renderResult = (
                <div className={className}>
                    {content}
                </div>
            );

        } else {
            var Content = this.state.content;
            renderResult = (
                <div className={className}>
                    <Content/>
                </div>
            );
        }

        return renderResult;
    }

    _HeaderComponent = () => {
        var rendeResult = null;
        if (this.state.title && this.state.title.length > 0) {
            rendeResult = (
                <header>{this.state.title}</header>
            );

        }
        return rendeResult;
    }

    _FooterComponent = (props) => {
        var classNames = ['dialog-button-block-primary'];

        if (this.state.buttons.length > 1) {
            classNames = [];
            for(var i=0; i < this.state.buttons.length -1; i++) {
                classNames.push('dialog-button-inline');
            }
            classNames.push('dialog-button-inline-primary')
        }

        var buttons = Object
            .keys(this.state.buttons)
            .map((key, i) => {
                var item = this.state.buttons[i];
                return (<DialogButton
                    key={i}
                    name={item.text}
                    className={classNames[i]
                    ? classNames[i]
                    : ""}
                    callback={item.callback}/>)
            })

        return (
            <div className="dialog-box-footer">{buttons}</div>
        )
    }

    render() {

        var dialogBoxClass = 'dialog-box';
        var dataAttr = {
            'data-dialog-name': this.state.dialogName
        };
        var hasExtraPaddingWithBody = false;

        switch (this.state.type) {
            default:
            case 0:
                return false;
            case 1:
            case 2:
                break;
            case 3:
                hasExtraPaddingWithBody = true;
                break;
            case 4:
                dialogBoxClass += ' dialog-list'
                break;
            case 5:
                dialogBoxClass += ' dialog-scene'
                break;
        }

        return (
            <div className="dialog" {...dataAttr}>
                <div className="dialog-backdrop"></div>
                <div className={dialogBoxClass}>
                    <div className="dialog-container">
                        <this._HeaderComponent ref={header => this.dialogHeader = header} dialogTitle={this.state.title}/>
                        <this._BodyComponent
                            ref={body => this.dialogBody = body}
                            extraPadding={hasExtraPaddingWithBody}/>
                        <this._FooterComponent ref={footer => this.dialogFooter = footer}/>
                    </div>
                </div>
            </div>
        )
    }
}

var MixiedDialog = DialogRelayoutMixin(Dialog);
export {MixiedDialog as default, Dialog as DialogCore,DialogStore, DialogActions};
