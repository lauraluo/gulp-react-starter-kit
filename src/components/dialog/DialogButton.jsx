import React, {Component} from 'react'
import {DialogActions} from "./Dialog"

class DialogButton extends Component {
    constructor(props) {
        super(props);
    }
    _onClick = () => {

        var isOverride = false;

        if (this.props.callback && typeof this.props.callback  == "function") {
            isOverride = this.props.callback();
        }

        if (isOverride && DialogActions) {
            DialogActions.hideDialog();
        }
    }

    render() {
        return (
            <div className={this.props.className} onClick={this._onClick}>{this.props.name}</div>
        );
    }
}

export default DialogButton;