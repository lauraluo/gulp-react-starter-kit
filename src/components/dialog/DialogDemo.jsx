import React from 'react'
import {DialogActions} from "./Dialog"
import $ from "jquery";

class DialogDemo extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            data: "appData"
        }
        this.subView = {};
    }

    _openCallback = () => {
        console.log("_openCallback");
        $(this.subView).find('input').focus();
    }
    
    _submitCallback = () => {
        console.log("submit");
        alert($(this.subView).find('input').val());
        return true;
    }
    
    _submitCallback2 = () => {
        console.log("submit");
        this.setState({data: "aadffd"});
        return true;
    }
    _subView = () => {
        return (
            <div ref={(subView)=> { this.subView = subView}}>
                <p className='title'>Sub view of DialogDemo APP abcd</p>
                <p>{this.state.data}</p>
                <input type="text"/>
                <button className='btn' onClick={this._submitCallback}>送出鈕</button>
            </div>
        )
    }

    _openDialog = () => {
        DialogActions.showDialog({
            title: "Dialog標題",
            content: this._subView,
            didOpened: this._openCallback,
            buttons:  [
                {
                    text: "取消",
                    callback: this._submitCallback2
                },
                {
                    text: "送出鈕",
                    callback: this._submitCallback
                }
            ]
        });
    }

    render() {
        return (
            <div className='dialogDemo'>
                <button className='btn' onClick={this._openDialog}>打開dialogavcd</button>
            </div >
        )
    }
}
export default DialogDemo