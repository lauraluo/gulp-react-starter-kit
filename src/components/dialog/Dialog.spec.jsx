//測試工具
import React from 'react'
import renderer from 'react-test-renderer'
import {mount, render, shallow} from 'enzyme'
import toJson from 'enzyme-to-json'

//被測試的相關模組
import DialogDemo from "./DialogDemo"
import Dialog from "./Dialog"
import Reflux from "reflux"
import {DialogActions} from "./Dialog"
import {DialogCore} from "./Dialog"
import DialogButton from "./DialogButton"


function getDialogInit(){
    return (<DialogCore ref={() => (null)} _relayout={() => (null)}/>);
}


describe('Dialog::', () => {
    describe('打開預設Dialog時，可以設定以下參數：標題，內容，開啟Dialog後的回調，按鈕(包含顯示名稱及點擊後的處理回調)', function () {
        let wrapper = mount(getDialogInit());
        let initState = wrapper.state();
        let mockOpenedCallback = jest.fn();
        let mockClickBtnCallback = [jest.fn(),jest.fn()];
        let mockContent = (<p>this is content</p>);
        let dialogConfig = {
            title: "DialogTitle1",
            content: function () {
                return mockContent
            },
            didOpened: mockOpenedCallback,
            buttons: [
                {
                    text: "取消",
                    callback: mockClickBtnCallback[0]
                }, {
                    text: "送出鈕",
                    callback: mockClickBtnCallback[1]
                }
            ]
        };

        beforeAll(()=>{
            DialogActions.showDialog(dialogConfig);
        });
        
        describe('呼叫showDialog成功後', function () {
            
            it("Dialog::state的isOpened", () => {
                expect(wrapper.state('isOpened')).toBe(true);
            })

            it("Dialog::的結構應該要出現在畫面上", () => {
                expect(wrapper.find('.dialog-backdrop').length).toBe(1);
                expect(wrapper.find('.dialog-container').length).toBe(1);
            })

            it("Dialog::應該要呼叫開啟dialog的回調", () => {
                expect(mockOpenedCallback).toHaveBeenCalledTimes(1);
            })


            it('Dialog::標題，為使用者設定的標題', () => {
                expect(wrapper.find('header').length).toEqual(1);                
                expect(wrapper.find('header').text()).toEqual(dialogConfig.title);
            });

            it('Dialog::內容，為使用者設定的內容', () => {
                expect(wrapper.contains(mockContent)).toEqual(true);
            });


            it('Dialog::按鈕，為使用者設定的按鈕', () => {
                expect(wrapper.find(DialogButton).length).toEqual(dialogConfig.buttons.length);

                for(var i=0; i< dialogConfig.buttons.length; i++){
                    expect(wrapper.find(DialogButton).at(i).props().name).toEqual(dialogConfig.buttons[i].text);                
                    expect(wrapper.find(DialogButton).at(i).props().callback ).toEqual(dialogConfig.buttons[i].callback);
                }             
                
                
            });

            it('Dialog::只有最後一個按鈕為主要按鈕', () => {
                expect(wrapper.find(DialogButton).last().props().className).toEqual('dialog-button-inline-primary');                
                expect(wrapper.find(DialogButton).first().props().className).not.toEqual('dialog-button-inline-primary');                
            });

            describe('當設定的按鈕被點擊時', function () {
                
                it('每一個按鈕的callback必需被調用',()=>{
                    wrapper.find(DialogButton).at(0).simulate('click');
                    expect(mockClickBtnCallback[0]).toHaveBeenCalledTimes(1);
                    wrapper.find(DialogButton).last().simulate('click');
                    expect(mockClickBtnCallback[1]).toHaveBeenCalledTimes(1);
                    
                });                

            });

            describe('使用者可以對Dialog的內容進行部份更新：允許更新的屬性有，標題，內容，按鈕', () => {
                    let newMockContent = (<p>this is new content</p>);
                    let updateDatas = {
                        title: "new Dialog title",
                        content: ()=> {
                            return newMockContent
                        },
                        buttons: [{
                            text: "新取消",
                            callback: jest.fn()
                        }, {
                            text: "新送出鈕",
                            callback: jest.fn()
                        }]
                    };


                    beforeAll(()=>{
                        DialogActions.updateDialog(updateDatas);
                    });

                    it('Dialog::標題，為使用者設定的標題', () => {
                        expect(wrapper.find('header').length).toEqual(1);                
                        expect(wrapper.find('header').text()).toEqual(updateDatas.title);
                    });

                    it('Dialog::內容，為使用者設定的內容', () => {
                        expect(wrapper.contains(newMockContent)).toEqual(true);
                    });


                    it('Dialog::按鈕，為使用者設定的按鈕', () => {
                        expect(wrapper.find(DialogButton).length).toEqual(updateDatas.buttons.length);

                        for(var i=0; i< updateDatas.buttons.length; i++){
                            expect(wrapper.find(DialogButton).at(i).props().name).toEqual(updateDatas.buttons[i].text);                
                            expect(wrapper.find(DialogButton).at(i).props().callback ).toEqual(updateDatas.buttons[i].callback);
                        }             
                    });

                    describe("如果title更新為空值",()=> {
                        beforeAll(()=>{
                            DialogActions.updateDialog({title: ''});
                        });

                        it('則header的結構必需消失', () => {
                            // console.log(wrapper.state());
                            expect(wrapper.find('header').length).toEqual(0);                
                        });  
                    });

                });


                describe('關閉Dialog', function () {

                    beforeAll(()=>{
                        DialogActions.hideDialog();
                    });

                    it('reflux的state必需回到初始值',()=>{
                        var closedState = wrapper.state();
                        Object.keys(closedState).map(function(key, index) {
                            if(initState[key]){
                                expect(closedState[key]).toEqual(initState[key]);  
                            }
                        });
                    });               
                });


        });



         

        // describe('Dialog打開方法的傳入參數，可以使用多型，包含以下內容', function () {
        //     describe('Dialog的Header：可以為空',function(){
        //         return false;
        //     });
            
        //     describe('Dialog內容：除了是functional component也可以是包含斷行符號的字串，畫面會用br取代斷行符號呈現字串',function(){
        //         return false;

        //     });
        // });
        
    });

    // describe('使用可以呼叫Confirm型的Dialog(目的在於簡化config的設定方式)',function(){
    //     describe('使用者只需要設定：Dialog標題、Dialog的內容、同意及取消的callback(陣列)', function () {
    //         return false;
    //     });
    //     describe('允許使用者自己客制同意及取消按鈕的文案', function () {
    //         return false;
    //     });          
    // });
});


// describe('如果正確安裝在doucment上', function () {
//     it('relayout應該要呼叫一次',()=>{

//     });
// });  