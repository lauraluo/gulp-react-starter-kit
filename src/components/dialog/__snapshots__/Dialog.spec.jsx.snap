// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog:: 打開預設Dialog時，可以設定以下參數：標題，內容，開啟Dialog後的回調，按鈕(包含顯示名稱及點擊後的處理回調) 呼叫showDialog成功後 Dialog::state的isOpened 1`] = `
<Dialog
  _relayout={[Function]}
>
  <div
    className="dialog"
  >
    <div
      className="dialog-backdrop"
    />
    <div
      className="dialog-box"
    >
      <div
        className="dialog-container"
      />
    </div>
  </div>
</Dialog>
`;
